cv_next/app/cv/[cvId]/components/commentSection/comment.tsx

```diff
@@ -1,78 +1,248 @@
   ) => void;
-}) {
+}
+
+export default function Comment({
+  comment,
+  userId,
+  childDepth = 0,
+  commentsOfComment = [],
+  setCommentsOfComments,
+}: CommentProps) {
   const [commentOnCommentStatus, setCommentOnCommentStatus] =
     useState<boolean>(false);
   const [commentOnComment, setCommentOnComment] = useState<string>("");
-  const router = useRouter();
   const [showAlert, setShowAlert] = useState<boolean>(false);
   const onAlertClick = async (type: boolean) => {
     if (type) await deleteCommentAction();
     setShowAlert(false);
   };
-  const supabase = useSupabase();
+  const { mutate } = useSWRConfig();
 
   const addNewCommentClickEvent = async () => {
-    const userId = (await supabase.auth.getUser()).data.user?.id;
-    if (!userId) router.push("/inactive");
-    else {
-      const commentToAdd: NewCommentModel = {
-        data: commentOnComment,
-        document_id: cv.id,
-        parent_comment_Id: comment.id,
-        user_id: userId,
-      };
+    const commentToAdd: NewCommentModel = {
+      data: commentOnComment,
+      document_id: comment.document_id,
+      parent_comment_Id: comment.id,
+      user_id: userId,
+    };
+
+    const optimisticData = (
+      prevData: CommentModel[] | undefined
+    ): CommentModel[] => {
+      const newCommentsOfComments = new Map<string, CommentModel[]>(
+        prevData?.reduce((acc, cur) => {
+          const key = cur.parent_comment_Id || "root"; // Use 'root' for top-level comments
+          if (!acc.has(key)) {
+            acc.set(key, []);
+          }
+          acc.get(key)!.push(cur);
+          return acc;
+        }, new Map<string, CommentModel[]>()) || []
+      );
+
+      const parentId = comment.id;
+      if (!newCommentsOfComments.has(parentId)) {
+        newCommentsOfComments.set(parentId, []);
+      }
+      newCommentsOfComments
+        .get(parentId)!
+        .push(commentToAdd as unknown as CommentModel);
 
-      await addComment(commentToAdd).finally(() => {
-        setCommentsOfComments((prev: Map<string, any[]>) => {
-          const newCommentsOfComments = new Map<string, any[]>(prev);
-          newCommentsOfComments.set(
-            comment.id,
-            newCommentsOfComments
-              .get(comment.id)
-              ?.concat([commentToAdd]) as Array<any>
-          );
-          return newCommentsOfComments;
-        });
-        setCommentOnCommentStatus(false);
+      return Array.from(newCommentsOfComments.values()).flat();
+    };
+
+    try {
+      mutate(comment.document_id, optimisticData, {
+        optimisticData: optimisticData(commentsOfComment),
+        rollbackOnError: (error) => {
+          // Type-checking for error
+          if (error instanceof Error) {
+            return error.name !== "AbortError";
+          }
+          // Rollback if error is not an instance of Error
+          return true;
+        },
       });
+      await addComment(commentToAdd);
+      mutate(comment.document_id);
```

Comment:
Why are you mutating twice? Is this how it's suppose to be handled? (I don't remember exactly the docs)

Reply:
First is to instantly show the data to the user, second is to make sure the cache is synced with the server's state

Reply:
So I believe according to docs that you should send the `addComment` function to the `mutate`, and it'll do the sync after the comment addition whilst optimistically updating the ui. Check the example here https://swr.vercel.app/docs/mutation#optimistic-updates. Also you might want to utilize (maybe in other places, not sure if here is needed) [manually updating the cache] (https://swr.vercel.app/docs/mutation#optimistic-updates).

Reply:
Took some time, but uploaded a fixed version